### DISCUSSION TITLE
Run each migration in a separate transaction

### DISCUSSION THREAD

**Discussion by pythonweb2** [sqlalchemy/alembic#1126] [Usage Questions] [answered]
Is there a way to run each migration in a seperate transaction? I am using Postgres, and I am trying to set up something like this with auto-migrations. 

script.py.mako
```python
"""${message}

Revision ID: ${up_revision}
Revises: ${down_revision | comma,n}
Create Date: ${create_date}

"""
from alembic import op
import sqlalchemy as sa
${imports if imports else ""}

# revision identifiers, used by Alembic.
revision = ${repr(up_revision)}
down_revision = ${repr(down_revision)}
branch_labels = ${repr(branch_labels)}
depends_on = ${repr(depends_on)}


def upgrade():
    with context.begin_transaction():
        ${upgrades if upgrades else "pass"}


def downgrade():
    with context.begin_transaction():
        ${downgrades if downgrades else "pass"}
```

env.py
```python
def run_migrations_online():
    """Run migrations in 'online' mode.

    In this scenario we need to create an Engine
    and associate a connection with the context.

    """
    connectable = engine_from_config(
        config.get_section(config.config_ini_section),
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    with connectable.connect() as connection:
        context.configure(connection=connection, target_metadata=target_metadata)
        # >>>> removed the transaction block here
        context.run_migrations()
```

A couple things. This feels a bit hacky, and I am just wondering if there is a better way to do it. Also, the mako template doesn't seem to add extra whitespace inside of the context manager.

example revision
```python
def upgrade():
    with context.begin_transaction():
        # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('transaction',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('amount', sa.Numeric(), nullable=True),
    sa.PrimaryKeyConstraint('id', name=op.f('pk_transaction'))
    )
    # ### end Alembic commands ###


def downgrade():
    with context.begin_transaction():
        # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('transaction')
    # ### end Alembic commands ###
```

**Comment by CaselIT** [comment]
hi,
Is using [`transaction_per_migration`](https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.configure.params.transaction_per_migration) not an option?

**Reply by pythonweb2** [reply]
Didn't know this existed, thank you!

**Comment by zzzeek** [comment, marked as answer]
> Is there a way to run each migration in a seperate transaction? I am using Postgres, and I am trying to set up something like this with auto-migrations.

absolutely, you can set [transaction_per_migration](https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.configure.params.transaction_per_migration) for this.

> 
> A couple things. This feels a bit hacky, and I am just wondering if there is a better way to do it. Also, the mako template doesn't seem to add extra whitespace inside of the context manager.

Right...indentation is not part of what it renders, if you wanted to do that you'd break it into separate methods.

**Reply by pythonweb2** [reply]
Awesome, I shouldn't need the extra white space with the `transaction_per_migration` parameter. Thanks so much for the quick response!
